### PR TITLE
Issue 028:  Revise Button behavior

### PR DESCRIPTION
**Pull request by boerge1**
see issue #28

### CODE REVIEW

**Comment by jochenberger** [comment]
LGTM on AiO, 3 buttons, `!invertVolumeButtons`

File: src/buttons.hpp [hunk 1]
```diff
@@ -6,10 +6,40 @@
...
+/*
+three Button
+============
+                                                      vol/next-prev inverted
+           admin-menu  idle        pause       play        play
+-------------------------------------------------------------------------
+pause      select                  play        pause       pause
+pauseLong  end adm     shortcut    shortcut    track#      track#
+up         next                                next        vol++
+upLong     next 10     shortcut    shortcut    vol++       next
```
Comment (jochenberger) [inline src/buttons.hpp:18]: Actually, upLong with inverted buttons is implemented as next, *cont.* now. Is that what you wanted?

File: src/buttons.hpp [hunk 2]
```diff
@@ -6,10 +6,40 @@
...
+down       prev                                prev        vol--
+downLong   prev 10     shortcut    shortcut    vol--       prev
```
Comment (jochenberger) [inline src/buttons.hpp:20]: Same here, downLong with inverted buttons is implemented as pref, *cont.*.

**Comment by jochenberger** [comment]
I recently started to wonder whether the continuous/non-continuous decision should really be made in the button layer. Maybe the command layer would be more appropriate.
Also, it's slightly awkward that the state machine needs to interpret the commands based on its state. E.g. `track` has different meanings depending on whether the box is playing or idle. I think that there should be dedicated commands for the shortcuts. But then, the command layer would have to know about the states and that would be a leaky abstraction as well. What do you think?

**Comment by jochenberger** [comment]
On the other hand, it would make sense, since for example, a potential shortcutN command should probably be triggered once whereas the volume up command should (IMHO) be triggered multiple times when the same button is held.
Oh, this is all so hard to write down. Maybe we should talk. 😉

**Comment by boerge1** [comment]
Yes, I'm not happy with the current solution. 
At the moment we have a problem with the 3-Button variant if pressing long for a shortcut. The shortcut command will then be repeated if pressed longer than 2 seconds. 
Have to think about that.

**Comment by jochenberger** [comment]
On the other hand, it would make sense, since for example, a potential shortcutN command should probably be triggered once whereas the volume up command should (IMHO) be triggered multiple times when the same button is held.
Oh, this is all so hard to write down. Maybe we should talk. 😉

**Comment by boerge1** [comment]
Ok, find now a better solution. The repeat selection is now implemented in the command class. 
Not fully tested but already pushed. 
@jochenberger Could you also help testing

**Comment by jochenberger** [comment]
I tested it on my 3-button `invertVolumeButtons` AiO and it works as expected. :+1:

File: src/buttons.hpp (hunk 1 above)
Comment (boerge1) [reply src/buttons.hpp]: I think now it is implemented only as next, without repetition

File: src/buttons.hpp (hunk 2 above)
Comment (boerge1) [reply src/buttons.hpp]: Same answer

**Comment by boerge1** [comment]
I tested AiO+ with 5 buttons. All as expected. 
So, I'm going to merge this PR in the next days.

**Comment by jochenberger** [comment]
I actually also shut down via longPause in my fork, so I'm glad you're adding that upstream.
However, I do it via a dedicated state that calls tonuino.shutdown() in its entry phase. I find that a little bit cleaner, but that's probably a matter of taste.

**Comment by jochenberger** [comment]
The discussion about the button usage made me wonder if they should be more freely configurable. Some people may rather want the shortcut to be played with pauseLong while others might like to start the toddler mode with a long pause press. Or long down for that matter.
I also still think that the shortcut should be dedicated commands. That would make the code cleaner IMHO.

File: src/state_machine.hpp [hunk 3]
```diff
@@ -82,6 +82,7 @@ class Base: public SM_tonuino
   bool readCard();
   bool handleShortcut(uint8_t shortCut);
   void handleReadCard();
+  bool checkRawForSortcutAndShutdown(commandRaw cmd_raw);
```
Comment (jochenberger) [inline src/state_machine.hpp:85]: typo "Sortcut"

File: src/state_machine.cpp [hunk 4]
```diff
@@ -651,6 +663,26 @@ void Base::handleReadCard() {
...
 }
 
+bool Base::checkRawForSortcutAndShutdown(commandRaw cmd_raw) {
+  uint8_t shortCut = 0xff;
+  switch(cmd_raw) {
+  case commandRaw::updownLong: shortCut = 0      ; break;
+  case commandRaw::upLong    : shortCut = 1      ; break;
+  case commandRaw::downLong  : shortCut = 2      ; break;
+  case commandRaw::start     : shortCut = 3      ; break;
+  case commandRaw::pauseLong : tonuino.shutdown(); break;
```
Comment (jochenberger) [inline src/state_machine.cpp:673]: Maybe play a jingle when shutting down manually?

Comment (boerge1) [reply src/state_machine.cpp:673]: Added in shutdown function. Maybe later replaced by "bye bye".

**Comment by boerge1** [comment]
> The discussion about the button usage made me wonder if they should be more freely configurable. Some people may rather want the shortcut to be played with pauseLong while others might like to start the toddler mode with a long pause press. Or long down for that matter. I also still think that the shortcut should be dedicated commands. That would make the code cleaner IMHO.

Yes, I did it in my last commit. See commands.cpp in the very beginning.

File: src/state_machine.cpp (hunk 4 above)
Comment (jochenberger) [reply src/state_machine.cpp]: Great! Should that also be played when the standby timer triggers the shutdown? I think it shouldn't.

**Comment by jochenberger** [comment]
I rebased my fork on your latest changes. LGTM

**Pull request merged**